src/modules/firewall/src/lib/Firewall.cpp

```diff
@@ -196,4 +679,309 @@ std::string IpTables::Fingerprint() const
+    {
+        if (IsFullLoggingEnabled())
+        {
+            OsConfigLogError(FirewallLog::Get(), "Failed to remove rule (%s): %s", command.c_str(), textResult);
+        }
+        else
+        {
+            OsConfigLogError(FirewallLog::Get(), "Failed to remove rule: %s", textResult);
+        }
+
+        error = textResult;
+    }
+
+    FREE_MEMORY(textResult);
+
+    return status;
+}
+
+int IpTables::SetRules(const std::vector<IpTables::Rule>& rules)
+{
+    int status = 0;
+    int index = rules.size() - 1;
+    std::string error;
+    std::vector<std::string> errors;
+
+    // Iterate through the rules in reverse order to ensure that the resulting
+    // rule order in the iptables chain is the same as the desired order
+    for (auto it = rules.rbegin(); it != rules.rend(); ++it, --index)
+    {
+        Rule rule = *it;
+
+        if (rule.HasParseError())
+        {
+            for (const std::string& parseError : rule.GetParseError())
+            {
+                errors.push_back("[" + std::to_string(index) + "] " + parseError);
+            }
+        }
+        else
+        {
+            switch (rule.GetDesiredState())
+            {
+                case Rule::State::Present:
+                    if (!Exists(rule))
+                    {
+                        if (0 != Add(rule, error))
+                        {
+                            errors.push_back("Failed to add rule (" + std::to_string(index) + "): " + error);
+                        }
+                    }
+                    break;
+
+                case Rule::State::Absent:
+                    while (Exists(rule))
+                    {
+                        if (0 != Remove(rule, error))
+                        {
+                            errors.push_back("Failed to remove rule (" + std::to_string(index) + "): " + error);
+                        }
+                    }
+                    break;
+
+                default:
+                    OsConfigLogError(FirewallLog::Get(), "Invalid desired rule state (%d): %d", index, static_cast<int>(rule.GetDesiredState()));
+                    status = EINVAL;
+            }
+        }
+    }
+
+    if (errors.size() > 0)
+    {
+        // Errors are in reverse order, so reverse them back to normal order
+        // and reset the status message/status code if there were any errors
+        std::reverse(errors.begin(), errors.end());
+        std::string errorMessage = "";
+
+        for (const std::string& error : errors)
+        {
+            errorMessage += error + "\n";
+        }
+
+        m_ruleStatusMessage = errorMessage;
+        status = -1;
+    }
+    else
+    {
+        m_policyStatusMessage = "";
+    }
+
+    return status;
+}
+
+std::vector<IpTablesPolicy> IpTables::GetDefaultPolicies() const
+{
+    const char* command = "iptables -S | grep -E '^-P (INPUT|OUTPUT)'";
+    static const std::regex policyRegex("^-P\\s+(\\S+)\\s+(\\S+)$");
```

Comment:
does this work across all our known distros? I suggest adding a recipe that exercises this code

Reply:
Yes, the firewall module already has a dependency on `iptables` (for fingerprint and enabled state) so the same will work for default policies etc.

Reply:
Agreed, I will add test recipes to this PR